src/analyzer/analyzerthread.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <chrono>
+#include <vector>
+
+#include "util/workerthread.h"
+
+#include "analyzer/analyzerprogress.h"
+#include "analyzer/analyzer.h"
+#include "control/controlvalue.h"
+#include "preferences/usersettings.h"
+#include "sources/audiosource.h"
+#include "track/track.h"
+#include "util/db/dbconnectionpool.h"
+#include "util/samplebuffer.h"
+#include "util/memory.h"
+
+
+enum class AnalyzerMode {
+    Default,
+    WithBeats,
+    WithBeatsWithoutWaveform,
+};
+
+enum class AnalyzerThreadState {
+    Void,
+    Idle,
+    Busy,
+    Done,
+    Exit,
+};
+
+Q_DECLARE_TYPEINFO(AnalyzerThreadState, Q_MOVABLE_TYPE);
+Q_DECLARE_METATYPE(AnalyzerThreadState);
+
+// Atomic control values are used for transferring data between the
+// host and the worker thread, e.g. the next track to be analyzed or
+// the current analyzer progress that can be read independent of any
+// progress signal
+//
+// The frequency of progress signal is limited to avoid flooding the
+// signal queued connection between the internal worker thread and
+// the host, which might otherwise cause unresponsiveness of the host.
+class AnalyzerThread : public WorkerThread {
+    Q_OBJECT
+
+  public:
+    AnalyzerThread(
+            int id,
+            mixxx::DbConnectionPoolPtr pDbConnectionPool,
+            UserSettingsPointer pConfig,
+            AnalyzerMode mode = AnalyzerMode::Default);
+    ~AnalyzerThread() override = default;
+
+    int id() const {
+        return m_id;
+    }
+
+    // Submits the next track to the worker thread without
+    // blocking. This is only allowed after a progress() signal
+    // with state Idle has been received to avoid overwriting
+    // a previously sent track that has not been received by the
+    // worker thread, yet.
+    void submitNextTrack(TrackPointer nextTrack);
+
+  signals:
+    // Use a single signal for progress updates to ensure that all signals
+    // are queued and received in the same order as emitted from the internal
+    // worker thread. Different signals would otherwise end up in different
+    // queued connections which are processed in an undefined order!
+    // TODO(uklotzde): Encapsulate all signal parameters into an
+    // AnalyzerThreadProgress object and register it as a new meta type.
+    void progress(int threadId, AnalyzerThreadState threadState, TrackId trackId, AnalyzerProgress trackProgress);
+
+  protected:
+    void exec() override;
+
+    FetchWorkResult fetchWork() override;
+
+  private:
+    /////////////////////////////////////////////////////////////////////////
+    // Immutable values and pointers (objects are thread-safe)
+    const int m_id;
+    const mixxx::DbConnectionPoolPtr m_pDbConnectionPool;
+    const UserSettingsPointer m_pConfig;
+    const AnalyzerMode m_mode;
+
+    /////////////////////////////////////////////////////////////////////////
+    // Thread-safe atomic values
+
+    ControlValueAtomic<TrackPointer> m_nextTrack;
```

Comment:
I am afraid this works as an unwanted TrackPointer cache, with 8 recent tracks. (The ring buffer size) 
It is atomic if we do not access the value with more than 8 threads. If you access it simultaneously with more threads, it uses a spin lock.
Can't we use the Track ID here and fetch it new from the cache once  it is scheduled? The track ID is a natural atomic.

Reply:
Agreed. With the fixed and improved track cache we don't need to keep the track pointers alive. References can now safely be dropped in worker threads.

Reply:
Oh, no, not possible. The queued track still needs to be loaded from the TrackDAO on the main tread.

Reply:
Just a few cached tracks is at least better than hundreds or thousands as currently. I'll make the ring buffer size adjustable to prevent spin locks.